### PR TITLE
Add writing compressed h5ad files

### PR DESCRIPTION
**Pull request by LouiseDck**

### CODE REVIEW

File: R/write_h5ad_helpers.R [hunk 1]
```diff
@@ -241,12 +256,14 @@ write_h5ad_string_array <- function(value, file, name, version = "0.2.0") {
...
+#' `"NONE"`, `"GZIP"` or `"LZF"`.
 #' @param version Encoding version of the element to write
-write_h5ad_categorical <- function(value, file, name, version = "0.2.0") {
+write_h5ad_categorical <- function(value, file, name, compression, version = "0.2.0") {
   rhdf5::h5createGroup(file, name)
-  rhdf5::h5write(levels(value), file, paste0(name, "/categories"))
-  rhdf5::h5write(as.integer(value) - 1, file, paste0(name, "/codes"))
-  rhdf5::h5write(is.ordered(value), file, paste0(name, "/ordered"))
+  hdf5_write_compressed(file, paste0(name, "/categories"), levels(value), compression)
+  hdf5_write_compressed(file, paste0(name, "/codes"), as.integer(value), compression)
```
Comment (ivirshup) [inline R/write_h5ad_helpers.R:265]: Do you still need the -1 here?

File: R/HDF5AnnData.R [hunk 2]
```diff
@@ -372,7 +386,7 @@ HDF5AnnData <- R6::R6Class("HDF5AnnData", # nolint
 #' to_HDF5AnnData(ad, "test.h5ad")
 #' # remove file
 #' file.remove("test.h5ad")
-to_HDF5AnnData <- function(adata, file) { # nolint
+to_HDF5AnnData <- function(adata, file, ...) { # nolint
```
Comment (rcannood) [inline R/HDF5AnnData.R:389]: See above

```suggestion
to_HDF5AnnData <- function(adata, file, compression = c("none", "gzip", "lzf")) { # nolint
```

File: R/HDF5AnnData.R [hunk 3]
```diff
@@ -387,6 +401,7 @@ to_HDF5AnnData <- function(adata, file) { # nolint
     var_names = adata$var_names,
     layers = adata$layers,
     obsp = adata$obsp,
-    varp = adata$varp
+    varp = adata$varp,
+    ...
```
Comment (rcannood) [inline R/HDF5AnnData.R:405]: See above

```suggestion
    compression = compression
```

File: R/write_h5ad.R [hunk 4]
```diff
@@ -5,6 +5,7 @@
 #' @param object The object to write, either a "SingleCellExperiment" or a
 #' "Seurat" object
 #' @param path Path of the file to write to
+#' @param ... Extra arguments will be passed to the appropriate convertor
```
Comment (rcannood) [inline R/write_h5ad.R:8]: See above

```suggestion
  #' @param compression The compression algorithm to use when writing the
  #'  HDF5 file. Can be one of `"none"`, `"gzip"` or `"lzf"`. Defaults to
  #' `"none"`.
```

File: R/write_h5ad.R [hunk 5]
```diff
@@ -62,21 +63,23 @@
 #'   #
 #'   # write_h5ad(obj, h5ad_file)
 #' }
-write_h5ad <- function(object, path) {
+write_h5ad <- function(object, path, ...) {
```
Comment (rcannood) [inline R/write_h5ad.R:66]: See above

```suggestion
write_h5ad <- function(object, path, compression = c("none", "gzip", "lzf")) {
```

File: R/write_h5ad.R [hunk 6]
```diff
@@ -62,21 +63,23 @@
...
   if (inherits(object, "SingleCellExperiment")) {
     from_SingleCellExperiment(
       object,
       output_class = "HDF5AnnData",
-      file = path
+      file = path,
+      ...
```
Comment (rcannood) [inline R/write_h5ad.R:72]: See above
```suggestion
      compression = compression
```

File: R/write_h5ad.R [hunk 7]
```diff
@@ -62,21 +63,23 @@
...
     )
   } else if (inherits(object, "Seurat")) {
     from_Seurat(
       object,
       output_class = "HDF5AnnData",
-      file = path
+      file = path,
+      ...
```
Comment (rcannood) [inline R/write_h5ad.R:79]: See above
```suggestion
      compression = compression
```

File: R/write_h5ad.R [hunk 8]
```diff
@@ -62,21 +63,23 @@
...
     )
   } else if (inherits(object, "AbstractAnnData")) {
-    to_HDF5AnnData(object, path)
+    to_HDF5AnnData(object, path, ...)
```
Comment (rcannood) [inline R/write_h5ad.R:82]: See above
```suggestion
    to_HDF5AnnData(object, path, compression = compression)
```

File: R/write_h5ad_helpers.R [hunk 9]
```diff
@@ -5,6 +5,8 @@
 #' @param value The value to write
 #' @param file Path to a H5AD file or an open H5AD handle
 #' @param name Name of the element within the H5AD file
+#' @param compression The compression to use when writing the element. Can be
+#' one of `"NONE"`, `"GZIP"` or `"LZF"`. Defaults to `"NONE"`.
```
Comment (rcannood) [inline R/write_h5ad_helpers.R:9]: See above

```suggestion
#' one of `"none"`, `"gzip"` or `"lzf"`. Defaults to `"none"`.
```

File: R/write_h5ad_helpers.R [hunk 10]
```diff
@@ -13,7 +15,8 @@
 #' `write_h5ad_element()` should always be used instead of any of the specific
 #' writing functions as it contains additional boilerplate to make sure
 #' elements are written correctly.
-write_h5ad_element <- function(value, file, name, ...) { # nolint
+write_h5ad_element <- function(value, file, name, compression = "NONE", ...) { # nolint
+  compression <- match.arg(compression, choices = c("GZIP", "LZF", "NONE"))
```
Comment (rcannood) [inline R/write_h5ad_helpers.R:19]: See above.

```suggestion
write_h5ad_element <- function(value, file, name, compression = c("none", "gzip", "lzf"), ...) { # nolint
  compression <- match.arg(compression)

File: R/write_h5ad_helpers.R [hunk 11]
```diff
@@ -101,8 +104,10 @@ write_h5ad_encoding <- function(file, name, encoding, version) {
 #' @param value Value to write
 #' @param file Path to a H5AD file or an open H5AD handle
 #' @param name Name of the element within the H5AD file
+#' @param compression The compression to use when writing the element. Can be
+#' one of `"NONE"`, `"GZIP"` or `"LZF"`.
```
Comment (rcannood) [inline R/write_h5ad_helpers.R:108]: See above

```suggestion
#' one of `"none"`, `"gzip"` or `"lzf"`. Defaults to `"none"`.
```

File: R/write_h5ad_helpers.R [hunk 12]
```diff
@@ -126,8 +131,10 @@ write_h5ad_dense_array <- function(value, file, name, version = "0.2.0") {
 #' @param value Value to write
 #' @param file Path to a H5AD file or an open H5AD handle
 #' @param name Name of the element within the H5AD file
+#' @param compression The compression to use when writing the element. Can be one of
+#' `"NONE"`, `"GZIP"` or `"LZF"`.
```
Comment (rcannood) [inline R/write_h5ad_helpers.R:135]: See above

```suggestion
#' @param compression The compression to use when writing the element. Can be
#' one of `"none"`, `"gzip"` or `"lzf"`. Defaults to `"none"`.
```

File: R/write_h5ad_helpers.R [hunk 13]
```diff
@@ -175,14 +182,17 @@ write_h5ad_sparse_array <- function(value, file, name, version = "0.1.0") {
...
```
Comment (rcannood) [inline R/write_h5ad_helpers.R:186]: See above

```suggestion
#' @param compression The compression to use when writing the element. Can be
#' one of `"none"`, `"gzip"` or `"lzf"`. Defaults to `"none"`.
```

File: R/write_h5ad_helpers.R [hunk 14]
```diff
@@ -197,14 +207,17 @@ write_h5ad_nullable_boolean <- function(value, file, name, version = "0.1.0") {
...
```
Comment (rcannood) [inline R/write_h5ad_helpers.R:211]: See above

```suggestion
#' @param compression The compression to use when writing the element. Can be
#' one of `"none"`, `"gzip"` or `"lzf"`. Defaults to `"none"`.
```

File: R/write_h5ad_helpers.R [hunk 15]
```diff
@@ -219,8 +232,10 @@ write_h5ad_nullable_integer <- function(value, file, name, version = "0.1.0") {
...
```
Comment (rcannood) [inline R/write_h5ad_helpers.R:236]: See above

```suggestion
#' @param compression The compression to use when writing the element. Can be
#' one of `"none"`, `"gzip"` or `"lzf"`. Defaults to `"none"`.
```

File: R/write_h5ad_helpers.R [hunk 16]
```diff
@@ -260,8 +277,10 @@ write_h5ad_categorical <- function(value, file, name, version = "0.2.0") {
...
```
Comment (rcannood) [inline R/write_h5ad_helpers.R:281]: See above

```suggestion
#' @param compression The compression to use when writing the element. Can be
#' one of `"none"`, `"gzip"` or `"lzf"`. Defaults to `"none"`.
```

File: tests/testthat/test-HDF5-write.R [hunk 17]
```diff
@@ -10,7 +10,7 @@ rhdf5::h5createFile(file = h5ad_file)
 test_that("Writing H5AD dense arrays works", {
   array <- matrix(rnorm(20), nrow = 5, ncol = 4)
 
-  expect_silent(write_h5ad_element(array, h5ad_file, "dense_array"))
+  expect_silent(write_h5ad_element(array, h5ad_file, "dense_array", compression = "NONE"))
```
Comment (rcannood) [inline tests/testthat/test-HDF5-write.R:13]:
```suggestion
  expect_silent(write_h5ad_element(array, h5ad_file, "dense_array", compression = "none"))

File: tests/testthat/test-HDF5-write.R [hunk 18]
```diff
@@ -21,7 +21,7 @@ test_that("Writing H5AD sparse arrays works", {
   array <- matrix(rnorm(20), nrow = 5, ncol = 4)
 
   csc_array <- as(array, "CsparseMatrix")
-  expect_silent(write_h5ad_element(csc_array, h5ad_file, "csc_array"))
+  expect_silent(write_h5ad_element(csc_array, h5ad_file, "csc_array", compression = "NONE"))
```
Comment (rcannood) [inline tests/testthat/test-HDF5-write.R:24]:
```suggestion
  expect_silent(write_h5ad_element(csc_array, h5ad_file, "csc_array", compression = "none"))

File: tests/testthat/test-HDF5-write.R [hunk 19]
```diff
@@ -31,7 +31,7 @@ test_that("Writing H5AD sparse arrays works", {
   expect_true(attrs[["encoding-type"]] == "csc_matrix")
 
   csr_array <- as(array, "RsparseMatrix")
-  expect_silent(write_h5ad_element(csr_array, h5ad_file, "csr_array"))
+  expect_silent(write_h5ad_element(csr_array, h5ad_file, "csr_array", compression = "NONE"))
```
Comment (rcannood) [inline tests/testthat/test-HDF5-write.R:34]:
```suggestion
  expect_silent(write_h5ad_element(csr_array, h5ad_file, "csr_array", compression = "none"))

File: tests/testthat/test-HDF5-write.R [hunk 20]
```diff
@@ -123,7 +123,7 @@ test_that("Writing H5AD mappings works", {
     scalar = 2
   )
 
-  expect_silent(write_h5ad_element(mapping, h5ad_file, "mapping"))
+  expect_silent(write_h5ad_element(mapping, h5ad_file, "mapping", compression = "NONE"))
```
Comment (rcannood) [inline tests/testthat/test-HDF5-write.R:126]:
```suggestion
  expect_silent(write_h5ad_element(mapping, h5ad_file, "mapping", compression = "NONE"))
```

File: tests/testthat/test-HDF5-write.R [hunk 21]
```diff
@@ -177,3 +177,45 @@ test_that("writing H5AD from Seurat works", {
...
+    var = dummy$var,
+    obs_names = dummy$obs_names,
+    var_names = dummy$var_names
+  )
+
+  h5ad_file_none <- tempfile(pattern = "hdf5_write_none_", fileext = ".h5ad")
+  h5ad_file_gzip <- tempfile(pattern = "hdf5_write_gzip_", fileext = ".h5ad")
+
+  write_h5ad(adata, h5ad_file_none, compression = "NONE")
+  write_h5ad(adata, h5ad_file_gzip, compression = "GZIP")
```
Comment (rcannood) [inline tests/testthat/test-HDF5-write.R:197]:
```suggestion
  write_h5ad(adata, h5ad_file_none, compression = "none")
  write_h5ad(adata, h5ad_file_gzip, compression = "gzip")
```

File: tests/testthat/test-HDF5-write.R [hunk 22]
```diff
@@ -177,3 +177,45 @@ test_that("writing H5AD from Seurat works", {
...
+    var = dummy$var,
+    obs_names = dummy$obs_names,
+    var_names = dummy$var_names
+  )
+
+  h5ad_file_none <- tempfile(pattern = "hdf5_write_none_", fileext = ".h5ad")
+  h5ad_file_lzf <- tempfile(pattern = "hdf5_write_lzf_", fileext = ".h5ad")
+
+  write_h5ad(adata, h5ad_file_none, compression = "NONE")
+  write_h5ad(adata, h5ad_file_lzf, compression = "LZF")
```
Comment (rcannood) [inline tests/testthat/test-HDF5-write.R:218]:
```suggestion
  write_h5ad(adata, h5ad_file_none, compression = "none")
  write_h5ad(adata, h5ad_file_lzf, compression = "lzf")

**Review by rcannood — CHANGES_REQUESTED** [review]
Woop woop! :partying_face: 

I made some minor suggestions -- WDYT @LouiseDck ?

About halfway through the review I started regretting not simply creating a PR with my proposed changes rather than all of these suggestions ...

**Review by rcannood — APPROVED** [review]

**Pull request merged**